HEATMAPS WITH PLACES API search function error

When I enter a location into the "HEATMAPS WITH PLACES API" demo, the app crashes and it gives the following message: "Unfortunately, Maps Utils Demo has stopped". The heatmap itself is not showing up either.

However, in the other "HEATMAPS DEMO ACTIVITY" the heatmap itself shows and all the other features (radius, gradient, etc) work completely fine. 

I created the secure.properties file and put my API key in the specified spot. I also entered my API key in the HeatMapsPlacesDemoActivity file where it says to. 


The device I am using is Moto G (3rd gen) and it is running android version 6.0.

I have not edited any files except adding in the secure.properties file and putting in my API key into the HeatMapsPlacesDemoActivity file as mentioned above. 

This is what the LogCat looks like once I enter in "Sydney" into the search bar and then click the "GO" button:


08-04 13:01:59.762 28667-28667/? I/art: Late-enabling -Xcheck:jni
08-04 13:01:59.854 28667-28667/com.google.maps.android.utils.demo W/System: ClassLoader referenced unknown path: /data/app/com.google.maps.android.utils.demo-2/lib/arm
08-04 13:01:59.903 28667-28667/com.google.maps.android.utils.demo W/art: Before Android 4.1, method android.graphics.PorterDuffColorFilter androidx.vectordrawable.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
08-04 13:01:59.950 28667-28667/com.google.maps.android.utils.demo I/art: Rejecting re-init on previously-failed class java.lang.Class<androidx.core.view.ViewCompat$2>
08-04 13:01:59.950 28667-28667/com.google.maps.android.utils.demo I/art: Rejecting re-init on previously-failed class java.lang.Class<androidx.core.view.ViewCompat$2>
08-04 13:02:00.111 28667-28689/com.google.maps.android.utils.demo D/OpenGLRenderer: Use EGL_SWAP_BEHAVIOR_PRESERVED: true
08-04 13:02:00.228 28667-28689/com.google.maps.android.utils.demo I/Adreno-EGL: <qeglDrvAPI_eglInitialize:379>: EGL 1.4 QUALCOMM build:  (I14a17c7611)
    OpenGL ES Shader Compiler Version: XE031.06.00.02
    Build Date: 12/04/15 Fri
    Local Branch: workspace
    Remote Branch: 
    Local Patches: 
    Reconstruct Branch: 
08-04 13:02:00.229 28667-28689/com.google.maps.android.utils.demo I/OpenGLRenderer: Initialized EGL, version 1.4
08-04 13:02:14.677 28667-28667/com.google.maps.android.utils.demo I/zzbz: Making Creator dynamically
08-04 13:02:14.690 28667-28667/com.google.maps.android.utils.demo I/DynamiteModule: Considering local module com.google.android.gms.maps_dynamite:0 and remote module com.google.android.gms.maps_dynamite:202414000
08-04 13:02:14.690 28667-28667/com.google.maps.android.utils.demo I/DynamiteModule: Selected remote version of com.google.android.gms.maps_dynamite, version >= 202414000
08-04 13:02:14.690 28667-28667/com.google.maps.android.utils.demo V/DynamiteModule: Dynamite loader version >= 2, using loadModule2NoCrashUtils
08-04 13:02:14.737 28667-28667/com.google.maps.android.utils.demo W/System: ClassLoader referenced unknown path: /data/data/com.google.android.gms/app_chimera/m/0000001c/n/armeabi-v7a
08-04 13:02:14.737 28667-28667/com.google.maps.android.utils.demo W/System: ClassLoader referenced unknown path: /data/data/com.google.android.gms/app_chimera/m/0000001c/n/armeabi
08-04 13:02:14.826 28667-28667/com.google.maps.android.utils.demo I/Google Maps Android API: Google Play services client version: 12451000
08-04 13:02:14.847 28667-28667/com.google.maps.android.utils.demo I/Google Maps Android API: Google Play services package version: 202414017
08-04 13:02:15.415 28667-28667/com.google.maps.android.utils.demo I/Choreographer: Skipped 54 frames!  The application may be doing too much work on its main thread.
08-04 13:02:17.401 28667-29099/com.google.maps.android.utils.demo W/DynamiteModule: Local module descriptor class for com.google.android.gms.googlecertificates not found.
08-04 13:02:17.432 28667-29099/com.google.maps.android.utils.demo I/DynamiteModule: Considering local module com.google.android.gms.googlecertificates:0 and remote module com.google.android.gms.googlecertificates:4
08-04 13:02:17.432 28667-29099/com.google.maps.android.utils.demo I/DynamiteModule: Selected remote version of com.google.android.gms.googlecertificates, version >= 4
08-04 13:02:17.451 28667-29099/com.google.maps.android.utils.demo W/System: ClassLoader referenced unknown path: /data/data/com.google.android.gms/app_chimera/m/0000001b/n/armeabi-v7a
08-04 13:02:17.451 28667-29099/com.google.maps.android.utils.demo W/System: ClassLoader referenced unknown path: /data/data/com.google.android.gms/app_chimera/m/0000001b/n/armeabi
08-04 13:02:32.675 28667-28678/com.google.maps.android.utils.demo I/art: Background partial concurrent mark sweep GC freed 15093(884KB) AllocSpace objects, 9(2MB) LOS objects, 25% free, 11MB/15MB, paused 8.373ms total 33.827ms

**_### **************THIS IS WHERE THE ERROR STARTS**************_**

08-04 13:02:32.786 28667-29149/com.google.maps.android.utils.demo E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #5
    Process: com.google.maps.android.utils.demo, PID: 28667
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:309)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
        at java.util.concurrent.FutureTask.run(FutureTask.java:242)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
        at java.lang.Thread.run(Thread.java:818)
     Caused by: java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare()
        at android.os.Handler.<init>(Handler.java:200)
        at android.os.Handler.<init>(Handler.java:114)
        at android.widget.Toast$TN.<init>(Toast.java:345)
        at android.widget.Toast.<init>(Toast.java:101)
        at android.widget.Toast.makeText(Toast.java:259)
        at com.google.maps.android.utils.demo.HeatmapsPlacesDemoActivity.getJsonPlaces(HeatmapsPlacesDemoActivity.java:265)
        at com.google.maps.android.utils.demo.HeatmapsPlacesDemoActivity.getPoints(HeatmapsPlacesDemoActivity.java:210)
        at com.google.maps.android.utils.demo.HeatmapsPlacesDemoActivity.access$200(HeatmapsPlacesDemoActivity.java:60)
        at com.google.maps.android.utils.demo.HeatmapsPlacesDemoActivity$MakeOverlayTask.doInBackground(HeatmapsPlacesDemoActivity.java:305)
        at com.google.maps.android.utils.demo.HeatmapsPlacesDemoActivity$MakeOverlayTask.doInBackground(HeatmapsPlacesDemoActivity.java:303)
        at android.os.AsyncTask$2.call(AsyncTask.java:295)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
        at java.lang.Thread.run(Thread.java:818) 
08-04 13:02:32.950 28667-28667/com.google.maps.android.utils.demo W/IInputConnectionWrapper: getTextAfterCursor on inactive InputConnection
08-04 13:02:33.028 28667-28689/com.google.maps.android.utils.demo E/Surface: getSlotFromBufferLocked: unknown buffer: 0xb9fc5898
08-04 13:02:33.028 28667-28689/com.google.maps.android.utils.demo D/OpenGLRenderer: endAllActiveAnimators on 0xb8ebeee8 (RippleDrawable) with handle 0xba042e70
08-04 13:02:33.037 28667-28667/com.google.maps.android.utils.demo W/IInputConnectionWrapper: beginBatchEdit on inactive InputConnection
08-04 13:02:33.037 28667-28667/com.google.maps.android.utils.demo W/IInputConnectionWrapper: endBatchEdit on inactive InputConnection
08-04 13:02:56.876 28667-29149/com.google.maps.android.utils.demo I/Process: Sending signal. PID: 28667 SIG: 9


Thanks for your help!

--- a/demo/src/main/java/com/google/maps/android/utils/demo/HeatmapsPlacesDemoActivity.java
+++ b/demo/src/main/java/com/google/maps/android/utils/demo/HeatmapsPlacesDemoActivity.java
@@ -19,6 +19,7 @@ package com.google.maps.android.utils.demo;
 import android.content.Context;
 import android.graphics.Color;
 import android.os.AsyncTask;
+import android.util.Log;
 import android.view.View;
 import android.view.inputmethod.EditorInfo;
 import android.view.inputmethod.InputMethodManager;
@@ -59,6 +60,8 @@ import java.util.Hashtable;
  */
 public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
 
+    private final String TAG = "HeatmapPlacesDemo";
+
     private final LatLng SYDNEY = new LatLng(-33.873651, 151.2058896);
 
     /**
@@ -69,7 +72,7 @@ public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
     /**
      * The options required for the radar search.
      */
-    private static final String TYPE_RADAR_SEARCH = "/radarsearch";
+    private static final String TYPE_NEARBY_SEARCH = "/nearbysearch";
     private static final String OUT_JSON = "/json";
 
     /**
@@ -194,7 +197,7 @@ public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
      *
      * @param keyword A string to use as a search term for the radar search
      * @return Returns the search results from radar search as a collection
-     * of LatLng objects.
+     * of LatLng objects, or null if there was an error calling the API
      */
     private Collection<LatLng> getPoints(String keyword) {
         HashMap<String, LatLng> results = new HashMap<>();
@@ -208,6 +211,10 @@ public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
 
         for (int j = 0; j < 4; j++) {
             String jsonResults = getJsonPlaces(keyword, searchCenters.get(j));
+            if (jsonResults == null) {
+                // Error calling Places API
+                return null;
+            }
             try {
                 // Create a JSON object hierarchy from the results
                 JSONObject jsonObj = new JSONObject(jsonResults);
@@ -215,16 +222,17 @@ public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
 
                 // Extract the Place descriptions from the results
                 for (int i = 0; i < pointsJsonArray.length(); i++) {
-                    if (!results.containsKey(pointsJsonArray.getJSONObject(i).getString("id"))) {
+                    if (!results.containsKey(pointsJsonArray.getJSONObject(i).getString("place_id"))) {
                         JSONObject location = pointsJsonArray.getJSONObject(i)
                                 .getJSONObject("geometry").getJSONObject("location");
-                        results.put(pointsJsonArray.getJSONObject(i).getString("id"),
+                        results.put(pointsJsonArray.getJSONObject(i).getString("place_id"),
                                 new LatLng(location.getDouble("lat"),
                                         location.getDouble("lng")));
                     }
                 }
             } catch (JSONException e) {
-                Toast.makeText(this, "Cannot process JSON results", Toast.LENGTH_SHORT).show();
+                Log.e(TAG, "Error parsing JSON:" + e);
+                runOnUiThread(() -> Toast.makeText(this, "Cannot process JSON results", Toast.LENGTH_SHORT).show());
             }
         }
         return results.values();
@@ -242,13 +250,14 @@ public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
         StringBuilder jsonResults = new StringBuilder();
         try {
             URL url = new URL(
-                    PLACES_API_BASE + TYPE_RADAR_SEARCH + OUT_JSON
-                    + "?location=" + location.latitude + "," + location.longitude
-                    + "&radius=" + (SEARCH_RADIUS / 2)
-                    + "&sensor=false"
-                    + "&key=" + API_KEY
-                    + "&keyword=" + keyword.replace(" ", "%20")
+                    PLACES_API_BASE + TYPE_NEARBY_SEARCH + OUT_JSON
+                            + "?location=" + location.latitude + "," + location.longitude
+                            + "&radius=" + (SEARCH_RADIUS / 2)
+                            + "&sensor=false"
+                            + "&key=" + API_KEY
+                            + "&keyword=" + keyword.replace(" ", "%20")
             );
+            Log.d(TAG, "URL: " + url);
             conn = (HttpURLConnection) url.openConnection();
             InputStreamReader in = new InputStreamReader(conn.getInputStream());
 
@@ -259,10 +268,10 @@ public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
                 jsonResults.append(buff, 0, read);
             }
         } catch (MalformedURLException e) {
-            Toast.makeText(this, "Error processing Places API URL", Toast.LENGTH_SHORT).show();
+            runOnUiThread(() -> Toast.makeText(this, "Error processing Places API URL", Toast.LENGTH_SHORT).show());
             return null;
         } catch (IOException e) {
-            Toast.makeText(this, "Error connecting to Places API", Toast.LENGTH_SHORT).show();
+            runOnUiThread(() -> Toast.makeText(this, "Error connecting to Places API", Toast.LENGTH_SHORT).show());
             return null;
         } finally {
             if (conn != null) {
@@ -302,10 +311,21 @@ public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
      */
     private class MakeOverlayTask extends AsyncTask<String, Integer, PointsKeywords> {
         protected PointsKeywords doInBackground(String... keyword) {
-            return new PointsKeywords(getPoints(keyword[0]), keyword[0]);
+            Collection<LatLng> points = getPoints(keyword[0]);
+            if (points != null) {
+                return new PointsKeywords(points, keyword[0]);
+            } else {
+                return null;
+            }
         }
 
         protected void onPostExecute(PointsKeywords pointsKeywords) {
+            ProgressBar progressBar = findViewById(R.id.progress_bar);
+            if (pointsKeywords == null) {
+                // Error calling Places API
+                progressBar.setVisibility(View.GONE);
+                return;
+            }
             Collection<LatLng> points = pointsKeywords.points;
             String keyword = pointsKeywords.keyword;
 
@@ -322,11 +342,9 @@ public class HeatmapsPlacesDemoActivity extends BaseDemoActivity {
                 }
                 mOverlaysRendered++;
                 if (mOverlaysRendered == mOverlaysInput) {
-                    ProgressBar progressBar = findViewById(R.id.progress_bar);
                     progressBar.setVisibility(View.GONE);
                 }
             } else {
-                ProgressBar progressBar = findViewById(R.id.progress_bar);
                 progressBar.setVisibility(View.GONE);
                 Toast.makeText(HeatmapsPlacesDemoActivity.this, "No results for this query :(", Toast.LENGTH_SHORT).show();
             }

